rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/CharacteristicLongWriteOperation.java

```diff
@@ -209,4 +215,52 @@ public Boolean call(Object emission) {
             }
         };
     }
+
+    private static Func1<Observable<? extends Throwable>, Observable<?>> errorIsRetryableAndAccordingTo(
+            final WriteOperationRetryStrategy writeOperationRetryStrategy,
+            final ByteBuffer byteBuffer,
+            final int batchSize) {
+        return new Func1<Observable<? extends Throwable>, Observable<?>>() {
+            @Override
+            public Observable<?> call(Observable<? extends Throwable> emittedOnWriteFailure) {
+                return emittedOnWriteFailure
+                        .flatMap(toLongWriteFailureOrError())
+                        .doOnNext(repositionByteBufferForRetry())
+                        .compose(writeOperationRetryStrategy);
+            }
+
+            @NonNull
+            private Func1<Throwable, Observable<WriteOperationRetryStrategy.LongWriteFailure>> toLongWriteFailureOrError() {
+                return new Func1<Throwable, Observable<WriteOperationRetryStrategy.LongWriteFailure>>() {
+                    @Override
+                    public Observable<WriteOperationRetryStrategy.LongWriteFailure> call(Throwable throwable) {
+                        if (!(throwable instanceof BleGattCharacteristicException || throwable instanceof BleGattCannotStartException)) {
+                            return Observable.error(throwable);
+                        }
+                        final int failedBatchNumber = calculateFailedBatchNumber(byteBuffer, batchSize);
+                        WriteOperationRetryStrategy.LongWriteFailure longWriteFailure = new WriteOperationRetryStrategy.LongWriteFailure(
+                                failedBatchNumber,
+                                (BleGattException) throwable
+                        );
+                        return Observable.just(longWriteFailure);
+                    }
+                };
+            }
+
+            @NonNull
+            private Action1<WriteOperationRetryStrategy.LongWriteFailure> repositionByteBufferForRetry() {
+                return new Action1<WriteOperationRetryStrategy.LongWriteFailure>() {
+                    @Override
+                    public void call(WriteOperationRetryStrategy.LongWriteFailure longWriteFailure) {
+                        final int newBufferPosition = longWriteFailure.getBatchNumber() * batchSize - batchSize;
+                        byteBuffer.position(newBufferPosition);
+                    }
+                };
+            }
+
+            private int calculateFailedBatchNumber(ByteBuffer byteBuffer, int batchSize) {
+                return (int) Math.ceil(byteBuffer.position() / (float) batchSize);
```

Comment:
doesn't it return a failed batch number + 1?

Reply:
Having a [1,2,3,4,5] and 2 bytes per batch
If the first batch fails the `byteBuffer.position()` would return 2
2/2 == 1 and the failed batch index should be 0

Reply:
Data size 5, batch size 2, failure at the last batch (3rd)
(int) Math.ceil(5 / (float) 2) => 3

Data size 6, batch size 2, failure at the last batch (3rd)
(int) Math.ceil(5 / (float) 2) => 3

Data size 5, batch size 3, failure at the last batch (2nd)
(int) Math.ceil(5 / (float) 2) => 2

Is this what you wanted to check?

Reply:
`calculateFailedBatchNumber` is misleading in this situation, either we should return
`(int) Math.ceil(byteBuffer.position() / (float) batchSize) - 1`
or name the function differently?

Reply:
It returns the "number" which I assumed starts from number one. `LongWriteFailure` also assumes "number". Do you think we should switch to zero-based indexes?

Reply:
Yup. This is user (developer) facing data
I think we should to zero-based indexes and rename the function to `calculateFailedBatchIndex` and `longWriteFailure.getBatchIndex()` respectively